src/workerd/api/messagechannel.c++

```diff
@@ -0,0 +1,317 @@
+#include "messagechannel.h"
+
+#include "global-scope.h"
+#include "worker-rpc.h"
+
+#include <workerd/io/worker.h>
+#include <workerd/jsg/ser.h>
+
+#include <capnp/message.h>
+
+namespace workerd::api {
+jsg::Ref<MessagePort::MessageEvent> MessagePort::MessageEvent::New(jsg::Lock& js,
+    jsg::JsRef<jsg::JsValue> data,
+    kj::String lastEventId,
+    jsg::Ref<MessagePort> source,
+    kj::Array<jsg::Ref<MessagePort>> ports) {
+  return jsg::alloc<MessagePort::MessageEvent>(
+      kj::str("message"), kj::mv(data), kj::mv(lastEventId), kj::mv(source), kj::mv(ports));
+}
+
+jsg::Ref<MessagePort::MessageEvent> MessagePort::MessageEvent::NewError(jsg::Lock& js,
+    jsg::JsRef<jsg::JsValue> data,
+    kj::String lastEventId,
+    jsg::Ref<MessagePort> source,
+    kj::Array<jsg::Ref<MessagePort>> ports) {
+  return jsg::alloc<MessagePort::MessageEvent>(
+      kj::str("messageerror"), kj::mv(data), kj::mv(lastEventId), kj::mv(source), kj::mv(ports));
+}
+
+MessagePort::MessagePort(IoContext& ioContext): ioContext(ioContext), state(Pending()) {
+  // We set a callback on the underlying EventTarget to be notified when
+  // a listener for the message event is added or removed. When there
+  // are no listeners, we move back to the Pending state, otherwise we
+  // will switch to the Started state if necessary.
+  setEventListenerCallback([&](jsg::Lock& js, kj::StringPtr name, size_t count) {
+    if (name == "message"_kj) {
+      KJ_SWITCH_ONEOF(state) {
+        KJ_CASE_ONEOF(pending, Pending) {
+          // If we are in the pending state, start the port if we have listeners.
+          // This is technically not spec compliant, but it is what Node.js
+          // supports. Specifically, adding a new message listener using the
+          // addEventListener method is *technically* not supposed to start
+          // the port but we're going to do what Node.js does.
```

Comment:
Why would we do what Node.js does and not what browsers do? Shouldn't we follow standards first?

Reply:
A key motivator for this is node.js compat... specifically, there are some modules/frameworks that we want to be able to support that are currently written to Node.js' implementation of `MessageChannel` that we would need to reconcile.